### PR TITLE
chore: adding more required functionality

### PR DESCRIPTION
**Pull request by priteshbandi**
* Adding more functionality required for notation-go to depend on this pkg.
* Also some renaming and update to Error struct to matching it with notation-go.

### CODE REVIEW

**Review by Two-Hearts — COMMENTED** [review]
1. Please add doc to struct `CLI` defined at line 23.
2. When I new a `CLI struct` with `pl == nil`: `cli, _ := New("notation-my-plugin", nil)`, this will `panic` when I call `cli.Execute(...)`. We should always check `nil` when new a `CLI`.
3. In `NewWithLogger` function, we are only checking prefix of the plugin executable name. We should also check the suffix matches with the plugin's `metadata.Name`, i.e. `{plugin-name}` of `notation-{plugin-name}` should equal to `pl.GetMetadata resp.Name`. In fact, do we really need the `executableName` parameter here?
4. `deferStdout()` and `type discardLogger struct{}` should be moved to `utils.go`
5. In `utils.go getValidArgs()`, the result strings should not be hardcoded, because they are defined in the `proto.go`: https://github.com/notaryproject/notation-plugin-framework-go/blob/c8c68d2fe13bd094d1bed57886d78a10d3ba5618/plugin/proto.go#L27

File: plugin/proto.go [hunk 1]
```diff
@@ -26,6 +32,11 @@ const (
 // Command is a CLI command available in the plugin contract.
 type Command string
 
+// Request defines a plugin request, which is always associated to a command.
+type Request interface {
+	Command() Command
+}
```
Comment (shizhMSFT) [inline plugin/proto.go:38]: What's the purpose of this interface?

Comment (priteshbandi) [reply plugin/proto.go:38]: Well, technically we shouldn't need this but the way its currently implemented in [notation-go](https://github.com/search?q=repo%3Anotaryproject%2Fnotation-go%20Command&type=code) we need this interface.

**Comment by priteshbandi** [comment]
> 1. Please add doc to struct `CLI` defined at line 23 of `cli.go`.
> 
>     2. When I new a `CLI struct` with `pl == nil`: `cli, _ := New("notation-my-plugin", nil)`, this will `panic` when I call `cli.Execute(...)`. We should always check `nil` when new a `CLI`.
> 
>     3. In `NewWithLogger` function, we are only checking prefix of the plugin executable name. We should also check the suffix matches with the plugin's `metadata.Name`, i.e. `{plugin-name}` of `notation-{plugin-name}` should equal to `pl.GetMetadata resp.Name`. In fact, do we really need the `executableName` parameter here?
> 
>     4. `deferStdout()` and `type discardLogger struct{}` should be moved to `utils.go`
> 
>     5. In `utils.go getValidArgs()`, the result strings should not be hardcoded, because they are defined in the `proto.go`: https://github.com/notaryproject/notation-plugin-framework-go/blob/c8c68d2fe13bd094d1bed57886d78a10d3ba5618/plugin/proto.go#L27

Done. Addressed all the feedbacks

File: plugin/proto.go [hunk 2]
```diff
@@ -26,6 +32,11 @@ const (
...
+
```
[the remark this answers is not in the document]
Comment (priteshbandi) [reply plugin/proto.go:39]: tracking it in https://github.com/notaryproject/notation-plugin-framework-go/pull/5#discussion_r1436200680

File: cli/cli.go [hunk 3]
```diff
@@ -19,24 +20,28 @@ import (
...
+	if pl == nil {
+		return nil, errors.New("plugin cannot be nil")
+	}
+
 	return &CLI{
-		name:   executableName,
 		pl:     pl,
 		logger: l,
-	}
+	}, nil
```
Comment (Two-Hearts) [inline cli/cli.go:44]: We have the `name` field defined in `CLI` struct. In `NewWithLogger`, we should get the plugin's metadata and fill `CLI.name` with `md.Name`.

**Review by Two-Hearts — COMMENTED** [review]
LGTM with a single nit comment.

File: cli/utils.go [hunk 4]
```diff
@@ -16,27 +17,89 @@ func getValidArgsString(md *plugin.GetMetadataResponse) string {
...
 	}
 
 	if slices.Contains(md.Capabilities, plugin.CapabilityTrustedIdentityVerifier) || slices.Contains(md.Capabilities, plugin.CapabilityRevocationCheckVerifier) {
-		opts = append(opts, "verify-signature")
+		opts = append(opts, plugin.CommandVerifySignature)
 	}
-	sort.Strings(opts)
-	return opts
+	// convert &opts (type *[]Command) to *[]string via unsafe.Pointer, then deref
+	stringOpts := *(*[]string)(unsafe.Pointer(&opts))
```
Comment (JeyJeyGao) [inline cli/utils.go:36]: Can we avoid using unsafe.Pointer here?
For example:
```go
opts := []string{
		string(plugin.CommandGetMetadata), string(plugin.Version),
	}
```

File: cli/cli.go [hunk 5]
```diff
@@ -19,24 +20,28 @@ import (
 	"github.com/notaryproject/notation-plugin-framework-go/plugin"
 )
 
+// CLI struct is used to create an executable for plugin.
 type CLI struct {
 	name   string
```
Comment (JeyJeyGao) [inline cli/cli.go:25]: Do we need the name field?

Comment (priteshbandi) [reply cli/cli.go]: Removed

File: cli/cli.go (hunk 3 above)
Comment (priteshbandi) [reply cli/cli.go:43]: Removed the field as its unused.

File: cli/utils.go (hunk 4 above)
Comment (priteshbandi) [reply cli/utils.go:36]: Yep that's possible but reduces readability and performance. Is ther4e any specific concern with  unsafe.Pointer as command is basically an alias for string?

**Review by Two-Hearts — APPROVED** [review]
LGTM

**Review by rgnote — APPROVED** [review]
LGTM

**Review by JeyJeyGao — APPROVED** [review]
LGTM

**Pull request merged**